common/codex32.c

```diff
@@ -13,6 +13,186 @@
+{
+	if (*x == 0 || y == 0) {
+		*x = 0;
+	} else {
+		*x = LOG_INV[(LOG[*x] + LOG[y]) % 31];
+	}
+	return;
+}
+
+/* Helper to input a single field element in the checksum engine. */
+static void input_fe(struct checksum_engine *checksum_engine, uint8_t e)
+{
+	size_t res_len = tal_count(checksum_engine->residue);
+	u8 xn = checksum_engine->residue[0];
+
+	for(size_t i = 1; i < res_len; i++) {
+		checksum_engine->residue[i - 1] = checksum_engine->residue[i];
+	}
+
+	checksum_engine->residue[res_len - 1] = e;
+
+	for(size_t i = 0; i < res_len; i++) {
+		u8 x = checksum_engine->generator[i];
+		multiply_gf32(&x, xn);
+		addition_gf32(&(checksum_engine->residue[i]), x);
+	}
+}
+
+/* Helper to input the HRP of codex32 string into the checksum engine */
+static void input_hrp(struct checksum_engine *codex_checksum_engine, const char *hrp)
+{
+	size_t i = 0;
+	for (i = 0; i < strlen(hrp); i++) {
+		input_fe(codex_checksum_engine, tolower(hrp[i]) >> 5);
+	}
+	input_fe(codex_checksum_engine, tolower(hrp[i]) >> 0);
+	for (i = 0; i < strlen(hrp); i++) {
+		input_fe(codex_checksum_engine, tolower(hrp[i]) & 0x1f);
+	}
+	return;
+}
+
+/* Helper to input data strong of codex32 into the checksum engine. */
+static void input_data_str(struct checksum_engine *codex_checksum_engine, const char *datastr)
+{
+	size_t i = 0;
+
+	for (i = 0; i < strlen(datastr); i++) {
+		input_fe(codex_checksum_engine, bech32_charset_rev[(int)datastr[i]]);
+	}
+
+	return;
+}
+
+/* Helper to verify codex32 checksum */
+static bool checksum_verify_short (const char *hrp, const char *codex_datastr)
+{
+	struct checksum_engine *codex_checksum_engine = tal(tmpctx, struct checksum_engine);
+
+	uint8_t generator[13] = {
+		bech32_charset_rev[(int)'E'], bech32_charset_rev[(int)'M'],
+		bech32_charset_rev[(int)'3'], bech32_charset_rev[(int)'G'],
+		bech32_charset_rev[(int)'Q'], bech32_charset_rev[(int)'E'],
+		bech32_charset_rev[(int)'E'], bech32_charset_rev[(int)'E'],
+		bech32_charset_rev[(int)'L'], bech32_charset_rev[(int)'M'],
+		bech32_charset_rev[(int)'C'], bech32_charset_rev[(int)'S'],
+		bech32_charset_rev[(int)'S']
+	};
+	uint8_t residue[13] = {
+		bech32_charset_rev[(int)'Q'], bech32_charset_rev[(int)'Q'],
+		bech32_charset_rev[(int)'Q'], bech32_charset_rev[(int)'Q'],
+		bech32_charset_rev[(int)'Q'], bech32_charset_rev[(int)'Q'],
+		bech32_charset_rev[(int)'Q'], bech32_charset_rev[(int)'Q'],
+		bech32_charset_rev[(int)'Q'], bech32_charset_rev[(int)'Q'],
+		bech32_charset_rev[(int)'Q'], bech32_charset_rev[(int)'Q'],
+		bech32_charset_rev[(int)'P']
+	};
+	uint8_t target[13] = {
+		bech32_charset_rev[(int)'S'], bech32_charset_rev[(int)'E'],
+		bech32_charset_rev[(int)'C'], bech32_charset_rev[(int)'R'],
+		bech32_charset_rev[(int)'E'], bech32_charset_rev[(int)'T'],
+		bech32_charset_rev[(int)'S'], bech32_charset_rev[(int)'H'],
+		bech32_charset_rev[(int)'A'], bech32_charset_rev[(int)'R'],
+		bech32_charset_rev[(int)'E'], bech32_charset_rev[(int)'3'],
+		bech32_charset_rev[(int)'2']
+	};
```

Comment:
Now, in this case you should make generator and target `static const u8` for maximum efficiency.

You don't need [13], since it's implied by the contents.
And you should use ARRAY_SIZE() (ccan/array_size/array_size.h) to hand them to input_hrp etc.